panic after GET: interface conversion: interface {} is nil, not string [recovered]

<!--
Thank you for helping to improve Crossplane!

Please be sure to search for open issues before raising a new one. We use issues
for bug reports and feature requests. Please find us at https://slack.crossplane.io
for questions, support, and discussion.
-->

I'm using Request to manage the lifecycle for the [registration of a GitLab Runner](https://docs.gitlab.com/runner/register/#register-with-a-runner-authentication-token) ([api](https://docs.gitlab.com/ee/api/users.html#create-a-runner-linked-to-a-user)).

The POST response payload returns a `token` which is used to register. The GET payload is different and does not include the token. The Warning is fair, that the token does not exist in the GET payload, but not sure where the panic is coming from.

In this use case, the Secret is created after the POST as expected. The Secret would never be updated as the token never expires or changes. When the runner is removed, that's when we want to de-register via DELETE.


### What happened?
<!--
Please let us know what behaviour you expected and how Crossplane diverged from
that behaviour.
-->
provider-http pod crashes with panic.


```
{"level":"info","ts":"2024-05-16T16:22:19Z","logger":"provider-http","msg":"http request sent: {\"method\":\"GET\",\"url\":\"https://***/api/v4/runners/42160\",\"headers\":{\"Content-Type\":[\"application/json\"],\"PRIVATE-TOKEN\":[\"***\"]}}","request":"example-runner"}
{"level":"info","ts":"2024-05-16T16:22:19Z","logger":"provider-http","msg":"Warning, value at field .body.token is empty, skipping secret update for: &{{\"id\":42160,\"description\":\"example runner\",\"ip_address\":null,\"active\":true,\"paused\":false,\"is_shared\":false,\"runner_type\":\"group_type\",\"name\":null,\"online\":null,\"status\":\"never_contacted\",\"tag_list\":[\"example-tags\"],\"run_untagged\":true,\"locked\":false,\"maximum_timeout\":null,\"access_level\":\"not_protected\",\"version\":null,\"revision\":null,\"platform\":null,\"architecture\":null,\"contacted_at\":null,\"projects\":[],\"groups\":[{\"id\":96,\"web_url\":\"https://***/groups/vt/si-devops\",\"name\":\"SI DevOps\"}]} map[Cache-Control:[max-age=0, private, must-revalidate] Connection:[keep-alive] Content-Length:[510] Content-Type:[application/json] Date:[Thu, 16 May 2024 16:22:20 GMT] Etag:[W/\"57c4b135a3bc04f5f7338dc7e237cf8a\"] Referrer-Policy:[strict-origin-when-cross-origin] Set-Cookie:[glbsticky=1715876541.321.17457.213499|155a34ed77d2269bfc4c9d6f5ac87e98; Expires=Thu, 16-May-24 16:27:20 GMT; Max-Age=300; Path=/; Secure; HttpOnly] Strict-Transport-Security:[max-age=63072000] Vary:[Origin] X-Content-Type-Options:[nosniff] X-Frame-Options:[SAMEORIGIN] X-Gitlab-Meta:[{\"correlation_id\":\"01HY13CBX29KRPP1QEKH3GKHXM\",\"version\":\"1\"}] X-Request-Id:[01HY13CBX29KRPP1QEKH3GKHXM] X-Runtime:[0.032985]] 200}","request":"example-runner"}
panic: interface conversion: interface {} is nil, not string [recovered]
        panic: interface conversion: interface {} is nil, not string

goroutine 143 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:119 +0x1a4
panic({0x11bb420, 0x40008596b0})
        runtime/panic.go:884 +0x1f4
github.com/crossplane-contrib/provider-http/internal/controller/request.(*external).desiredState(...)
        github.com/crossplane-contrib/provider-http/internal/controller/request/observe.go:99
github.com/crossplane-contrib/provider-http/internal/controller/request.(*external).isUpToDate(0x4000629a70, {0x1604208, 0x40006296e0}, 0x4000175400)
        github.com/crossplane-contrib/provider-http/internal/controller/request/observe.go:62 +0x60c
github.com/crossplane-contrib/provider-http/internal/controller/request.(*external).Observe(0x4000629a70, {0x1604208, 0x40006296e0}, {0x1617d88?, 0x4000175400})
        github.com/crossplane-contrib/provider-http/internal/controller/request/request.go:146 +0x60
github.com/crossplane/crossplane-runtime/pkg/reconciler/managed.(*Reconciler).Reconcile(0x4000184840, {0x1604240, 0x4000629620}, {{{0x0, 0x0}, {0x40000c9728, 0x12}}})
        github.com/crossplane/crossplane-runtime@v0.20.0-rc.0.0.20230413174155-c8cff1a7fb74/pkg/reconciler/managed/reconciler.go:805 +0x21f0
github.com/crossplane/crossplane-runtime/pkg/ratelimiter.(*Reconciler).Reconcile(0x40000888c0, {0x1604240, 0x4000629620}, {{{0x0?, 0x30?}, {0x40000c9728?, 0xffff69e06701?}}})
        github.com/crossplane/crossplane-runtime@v0.20.0-rc.0.0.20230413174155-c8cff1a7fb74/pkg/ratelimiter/reconciler.go:54 +0x124
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1604240?, {0x1604240?, 0x4000629620?}, {{{0x0?, 0x112c420?}, {0x40000c9728?, 0x15f1fd0?}}})
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:122 +0x8c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0x40004826e0, {0x1604198, 0x400007f810}, {0x1204e20?, 0x40004146c0?})
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:323 +0x2c0
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0x40004826e0, {0x1604198, 0x400007f810})
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274 +0x1a0
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235 +0x74
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:231 +0x43c
```


### How can we reproduce it?
<!--
Help us to reproduce your bug as succinctly and precisely as possible. Artifacts
such as example manifests or a script that triggers the issue are highly
appreciated!
-->

```
---
apiVersion: pkg.crossplane.io/v1
kind: Provider
metadata:
  name: provider-http
spec:
  package: "xpkg.upbound.io/crossplane-contrib/provider-http:v1.0.0"
---
apiVersion: http.crossplane.io/v1alpha1
kind: ProviderConfig
metadata:
  name: http-conf
spec:
  credentials:
    source: InjectedIdentity
---
apiVersion: http.crossplane.io/v1alpha2
kind: Request
metadata:
  name: example-runner
spec:
  forProvider:
    headers:
      PRIVATE-TOKEN:
        - "***"
      Content-Type:
        - application/json
    payload:
      baseUrl: "https://***/api/v4/"
      body: |-
        {
          "runner_type": "group_type",
          "group_id": 96,
          "description": "example runner",
          "tag_list": ["example-tags"]
          "paused": false,
          "run_untagged": false,
          "locked": false,
          "access_level": "not_protected",
          "maintenance_note": "",
        }
    mappings:
      - method: "POST"
        body: |
          {
            "runner_type": .payload.body.runner_type,
            "group_id": .payload.body.group_id,
            "description": .payload.body.description,
            "tag_list": .payload.body.tag_list,
            "run_untagged": .payload.body.run_untagged,
            "paused": .payload.body.paused,
            "locked": .payload.body.locked,
            "access_level": .payload.body.access_level
          }
        url: (.payload.baseUrl + "user/runners")
      - method: "GET"
        url: (.payload.baseUrl + "runners/" + (.response.body.id|tostring))
      - method: "DELETE"
        url: (.payload.baseUrl + "runners/" + (.response.body.id|tostring))
    secretInjectionConfigs:
      - secretRef:
          name: gitlab-runner-registration
          namespace: default
        secretKey: runner-token
        responsePath: .body.token
  providerConfigRef:
    name: http-conf
```


### What environment did it happen in?
Crossplane version: 1.15.3
provider-http: v1.0.0

rancher-desktop 1.13.1
kubernetes: 1.28.7

--- a/internal/controller/request/observe.go
+++ b/internal/controller/request/observe.go
@@ -2,6 +2,7 @@ package request
 
 import (
 	"context"
+	"fmt"
 	"net/http"
 	"strings"
 
@@ -61,6 +62,12 @@ func (c *external) isUpToDate(ctx context.Context, cr *v1alpha2.Request) (Observ
 	c.patchResponseToSecret(ctx, cr, &details.HttpResponse)
 	desiredState, err := c.desiredState(ctx, cr)
 	if err != nil {
+		if isErrorMappingNotFound(err) {
+			// Since there is no PUT mapping, we skip the check for its presence in the GET response.
+			return NewObserve(details, responseErr, true), nil
+		}
+
+		// For any other error, we return a failed observation.
 		return FailedObserve(), err
 	}
 
@@ -96,7 +103,11 @@ func (c *external) compareResponseAndDesiredState(details httpClient.HttpDetails
 
 func (c *external) desiredState(ctx context.Context, cr *v1alpha2.Request) (string, error) {
 	requestDetails, err := c.requestDetails(ctx, cr, http.MethodPut)
-	return requestDetails.Body.Encrypted.(string), err
+	if err != nil {
+		return "", err
+	}
+
+	return requestDetails.Body.Encrypted.(string), nil
 }
 
 func (c *external) requestDetails(ctx context.Context, cr *v1alpha2.Request, method string) (requestgen.RequestDetails, error) {
@@ -106,4 +117,10 @@ func (c *external) requestDetails(ctx context.Context, cr *v1alpha2.Request, met
 	}
 
 	return generateValidRequestDetails(ctx, c.localKube, cr, mapping)
+}
+
+// isErrorMappingNotFound checks if the provided error indicates that the
+// mapping for an HTTP PUT request is not found.
+func isErrorMappingNotFound(err error) bool {
+	return errors.Cause(err).Error() == fmt.Sprintf(errMappingNotFound, http.MethodPut)
 }

--- a/internal/controller/request/observe_test.go
+++ b/internal/controller/request/observe_test.go
@@ -144,6 +144,46 @@ func Test_isUpToDate(t *testing.T) {
 				},
 			},
 		},
+		"SuccessNoPUTMapping": {
+			args: args{
+				http: &MockHttpClient{
+					MockSendRequest: func(ctx context.Context, method string, url string, body, headers httpClient.Data, skipTLSVerify bool) (resp httpClient.HttpDetails, err error) {
+						return httpClient.HttpDetails{
+							HttpResponse: httpClient.HttpResponse{
+								Body:       `{"username":"old_name"}`,
+								StatusCode: 200,
+							},
+						}, nil
+					},
+				},
+				localKube: &test.MockClient{
+					MockStatusUpdate: test.NewMockSubResourceUpdateFn(nil),
+				},
+				mg: httpRequest(func(r *v1alpha2.Request) {
+					r.Status.Response.Body = `{"username":"john_doe_new_username"}`
+					r.Status.Response.StatusCode = 200
+					r.Spec.ForProvider.Mappings = []v1alpha2.Mapping{
+						testPostMapping,
+						testGetMapping,
+						testDeleteMapping,
+					}
+				}),
+			},
+			want: want{
+				err: nil,
+				result: ObserveRequestDetails{
+					Details: httpClient.HttpDetails{
+						HttpResponse: httpClient.HttpResponse{
+							Body:       `{"username":"old_name"}`,
+							Headers:    nil,
+							StatusCode: 200,
+						},
+					},
+					ResponseError: nil,
+					Synced:        true,
+				},
+			},
+		},
 		"SuccessJSONBody": {
 			args: args{
 				http: &MockHttpClient{

